### DISCUSSION TITLE
Unexpected behaviour with "colour.colorimetry.reshape_sd" definition

### DISCUSSION THREAD

**Discussion by frankkjensen** [colour-science/colour#1105] [Q&A] [answered]
I have an issue when reshaping some spectra. Below I load a spectrum with non-uniform wavelength intervals (file attached: [White CRI98 6000K.csv](https://github.com/colour-science/colour/files/10741618/White.CRI98.6000K.csv))

When reshaping to (400, 700, 5) and (300, 700, 5) it works as expected, but when I try (300, 700, 5) the step is 5.033333, except for one point, where it is 2.36666.

```
import colour

sds = colour.read_sds_from_csv_file('White CRI98 6000K.csv', delimiter=',', header=None, names=['Wavelength', 'White CRI98 6000K'])
sd = sds['White CRI98 6000K']

wls = sd.wavelengths
wlSteps = [wls[i+1] - wls[i] for i in range(len(wls)-1)]
print(wls)
print('Min:', min(wlSteps), '\tMax:', max(wlSteps), '\tAverage:', sum(wlSteps)/len(wlSteps))
# The wavelength distribution is not uniform
print(sd.shape)
# Shape reports the minimum wavelength interval

sd_reshaped = colour.colorimetry.reshape_sd(sd, colour.SpectralShape(400, 700, 5))
wls = sd_reshaped.wavelengths
wlSteps = [wls[i+1] - wls[i] for i in range(len(wls)-1)]
print(wls)
print('Min:', min(wlSteps), '\tMax:', max(wlSteps), '\tAverage:', sum(wlSteps)/len(wlSteps))
print(sd_reshaped.shape)
# Works as expected

sd_reshaped = colour.colorimetry.reshape_sd(sd, colour.SpectralShape(300, 700, 1))
wls = sd_reshaped.wavelengths
wlSteps = [wls[i+1] - wls[i] for i in range(len(wls)-1)]
print(wls)
print('Min:', min(wlSteps), '\tMax:', max(wlSteps), '\tAverage:', sum(wlSteps)/len(wlSteps))
print(sd_reshaped.shape)
# Works as expected

sd_reshaped = colour.colorimetry.reshape_sd(sd, colour.SpectralShape(300, 700, 5))
wls = sd_reshaped.wavelengths
wlSteps = [wls[i+1] - wls[i] for i in range(len(wls)-1)]
print(wls)
print('Min:', min(wlSteps), '\tMax:', max(wlSteps), '\tAverage:', sum(wlSteps)/len(wlSteps))
print(sd_reshaped.shape)
# The wavelength interval is 5.033333, except for one point.

```

Best regards,
Frank

**Comment by thomasmansencal** [comment, marked as answer]
Hi @frankkjensen,

The code tries to preserve as much as possible of the original data, there are some explanations [in this issue](https://github.com/colour-science/colour/issues/998) but let me give you some more details. All the magic happens in the [colour.SpectralDistribution.interpolate](https://github.com/colour-science/colour/blob/fb6f7aa466e2a9b6c3ea4e4ad98fedbf8b789eb0/colour/colorimetry/spectrum.py#L1155) method.

Given the requested shape is `(300, 700, 5)`, the code finds the minimum and maximum wavelength in the distribution of your spectrum: `(397.07, 779.67)`

- There is no data before 397.07, thus the next usable point for interpolation is 398
- There is no data beyond 779.67, thus the previous usable point for interpolation is 779
- We can thus safely interpolate between 398 and 779 on integer wavelengths
- The requested shape is `(300, 700, 5)` thus the effective interpolation shape becomes `(398, 700, 5)`

Such a shape will generate fractional intervals!

It is a case where I would recommend doing a direct extraction of the values you need as follows:

```python
sd[colour.SpectralShape(300, 700, 5).wavelengths]
```

Hope it makes sense!

Cheers,

Thomas

**Comment by frankkjensen** [comment]
Ok, I understand.
I was trying to convert several spectra with different wavelengths to the same shape to be able to convert them to a multi-spectral distribution, which requires the same shape (right?). But really 99.9% of the times, I would chose 1nm steps (or an integer faction of 1nm), and that would work just fine since the start and end are rounded to integers, right?

Best regards,
Frank
